File: backend/app/schemas/pos.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

# Till Schemas
class TillBase(BaseModel):
    till_number: str
    name: str
    location: Optional[str] = None
    hardware_config: Optional[dict] = None
    is_active: bool = True
    default_warehouse_id: int
    default_customer_id: Optional[int] = None
    branch_id: Optional[int] = None

class TillCreate(TillBase):
    pass

class TillUpdate(BaseModel):
    name: Optional[str] = None
    location: Optional[str] = None
    hardware_config: Optional[dict] = None
    is_active: Optional[bool] = None
    default_warehouse_id: Optional[int] = None
    default_customer_id: Optional[int] = None
    branch_id: Optional[int] = None

class Till(TillBase):
    id: int
    company_id: int
    
    class Config:
        from_attributes = True

# Till Session Schemas
class TillSessionOpen(BaseModel):
    till_id: int
    opening_balance: Decimal = Field(..., ge=0)

class POSSessionCreate(BaseModel):
    till_id: int
    opening_cash: Decimal = Field(..., ge=0)

class TillSessionClose(BaseModel):
    actual_closing_balance: Decimal = Field(..., ge=0)
    reconciliation_notes: Optional[str] = None

class TillReconciliationDetail(BaseModel):
    payment_method: str
    counted_amount: Decimal
    notes: Optional[str] = None

class TillSessionReconcile(BaseModel):
    reconciliation_details: List[TillReconciliationDetail]

class TillSession(BaseModel):
    id: int
    till_id: int
    user_id: int
    opening_date: datetime
    closing_date: Optional[datetime]
    opening_balance: Decimal
    expected_closing_balance: Optional[Decimal]
    actual_closing_balance: Optional[Decimal]
    variance: Optional[Decimal]
    status: str
    
    class Config:
        from_attributes = True

# POS Transaction Type Schemas
class POSTransactionTypeBase(BaseModel):
    name: str
    base_type: str
    default_payment_method: Optional[str] = None
    gl_account_id: Optional[int] = None
    is_active: bool = True

class POSTransactionTypeCreate(POSTransactionTypeBase):
    pass

class POSTransactionTypeUpdate(BaseModel):
    name: Optional[str] = None
    default_payment_method: Optional[str] = None
    gl_account_id: Optional[int] = None
    is_active: Optional[bool] = None

class POSTransactionType(POSTransactionTypeBase):
    id: int
    company_id: int
    
    class Config:
        from_attributes = True

# POS Transaction Line Schemas
class POSTransactionLineCreate(BaseModel):
    item_id: int
    quantity: Decimal = Field(..., gt=0)
    unit_price: Decimal
    discount_percentage: Decimal = Field(default=0, ge=0, le=100)
    discount_amount: Decimal = Field(default=0, ge=0)
    tax_type_id: Optional[int] = None

class POSTransactionLineRead(POSTransactionLineCreate):
    id: int
    description: str
    tax_amount: Decimal
    line_total: Decimal
    
    class Config:
        from_attributes = True

# POS Payment Schemas
class POSPaymentCreate(BaseModel):
    payment_method: str
    amount: Decimal = Field(..., gt=0)
    reference_number: Optional[str] = None
    payment_details: Optional[dict] = None

class POSPaymentRead(POSPaymentCreate):
    id: int
    
    class Config:
        from_attributes = True

# POS Transaction Schemas
class POSTransactionCreate(BaseModel):
    customer_id: Optional[int] = None
    transaction_type_id: int
    lines: List[POSTransactionLineCreate]
    payments: List[POSPaymentCreate]
    discount_amount: Decimal = Field(default=0, ge=0)

class POSTransactionRead(BaseModel):
    id: int
    transaction_number: str
    transaction_date: datetime
    customer_id: Optional[int]
    subtotal: Decimal
    tax_amount: Decimal
    discount_amount: Decimal
    total_amount: Decimal
    status: str
    lines: List[POSTransactionLineRead]
    payments: List[POSPaymentRead]
    
    class Config:
        from_attributes = True

class POSReturnCreate(BaseModel):
    original_transaction_id: int
    return_lines: List[POSTransactionLineCreate]
    reason: str

class POSCashMovementCreate(BaseModel):
    session_id: int
    movement_type: str  # "cash_in" or "cash_out"
    amount: Decimal = Field(..., gt=0)
    reason: str
    reference: Optional[str] = None

# POS Defaults Schemas
class POSDefaultsBase(BaseModel):
    default_walk_in_customer_id: Optional[int] = None
    receipt_header: Optional[str] = None
    receipt_footer: Optional[str] = None
    auto_print_receipt: bool = True
    allow_negative_stock: bool = False
    require_customer_for_credit: bool = True
    default_tax_type_id: Optional[int] = None
    cash_rounding_method: str = "None"
    cash_rounding_precision: Decimal = 0.01

class POSDefaultsCreate(POSDefaultsBase):
    pass

class POSDefaultsUpdate(POSDefaultsBase):
    pass

class POSDefaults(POSDefaultsBase):
    id: int
    company_id: int
    
    class Config:
        from_attributes = True

# Receipt Schema
class ReceiptData(BaseModel):
    transaction: POSTransactionRead
    company_info: dict
    till_info: dict
    cashier_name: str
    receipt_header: Optional[str]
    receipt_footer: Optional[str]

# Aliases for backward compatibility
POSTransaction = POSTransactionRead
POSTransactionLine = POSTransactionLineRead
POSPayment = POSPaymentRead
POSCashMovement = POSCashMovementCreate
POSSession = TillSession
POSSessionClose = TillSessionClose

File: frontend/src/types/pos.ts
export interface Till {
  id: number;
  company_id: number;
  till_code: string;
  name: string;
  warehouse_id: number;
  gl_cash_account_id: number;
  is_active: boolean;
  warehouse_name?: string;
  gl_cash_account_name?: string;
}

export interface TillCreate {
  till_code: string;
  name: string;
  warehouse_id: number;
  gl_cash_account_id: number;
  is_active?: boolean;
}

export interface TillUpdate {
  till_code?: string;
  name?: string;
  warehouse_id?: number;
  gl_cash_account_id?: number;
  is_active?: boolean;
}

export interface POSTransactionType {
  id: number;
  company_id: number;
  type_code: string;
  name: string;
  is_sale: boolean;
  is_return: boolean;
  gl_revenue_account_id?: number;
  gl_cost_account_id?: number;
  is_active: boolean;
  gl_revenue_account_name?: string;
  gl_cost_account_name?: string;
}

export interface POSTransactionTypeCreate {
  type_code: string;
  name: string;
  is_sale: boolean;
  is_return: boolean;
  gl_revenue_account_id?: number;
  gl_cost_account_id?: number;
  is_active?: boolean;
}

export interface POSTransactionTypeUpdate {
  type_code?: string;
  name?: string;
  is_sale?: boolean;
  is_return?: boolean;
  gl_revenue_account_id?: number;
  gl_cost_account_id?: number;
  is_active?: boolean;
}

export interface POSDefaults {
  id: number;
  company_id: number;
  default_warehouse_id?: number;
  default_customer_id?: number;
  default_sale_transaction_type_id?: number;
  default_return_transaction_type_id?: number;
  receipt_header?: string;
  receipt_footer?: string;
  auto_print_receipt: boolean;
  default_warehouse_name?: string;
  default_customer_name?: string;
  default_sale_transaction_type_name?: string;
  default_return_transaction_type_name?: string;
}

export interface POSDefaultsUpdate {
  default_warehouse_id?: number;
  default_customer_id?: number;
  default_sale_transaction_type_id?: number;
  default_return_transaction_type_id?: number;
  receipt_header?: string;
  receipt_footer?: string;
  auto_print_receipt?: boolean;
}

export interface POSSession {
  id: number;
  company_id: number;
  till_id: number;
  cashier_id: number;
  opening_cash_amount: number;
  closing_cash_amount?: number;
  expected_cash_amount?: number;
  cash_variance?: number;
  status: 'open' | 'closed';
  opened_at: string;
  closed_at?: string;
  notes?: string;
  till_name?: string;
  cashier_name?: string;
}

export interface POSSessionCreate {
  till_id: number;
  opening_cash_amount: number;
  notes?: string;
}

export interface POSSessionClose {
  closing_cash_amount: number;
  notes?: string;
}

export interface POSTransactionLine {
  id?: number;
  line_number: number;
  inventory_item_id: number;
  quantity: number;
  unit_price: number;
  discount_amount?: number;
  tax_amount?: number;
  line_total: number;
  inventory_item_name?: string;
  inventory_item_code?: string;
}

export interface POSTransaction {
  id: number;
  company_id: number;
  session_id: number;
  transaction_type_id: number;
  customer_id?: number;
  subtotal_amount: number;
  discount_amount?: number;
  tax_amount?: number;
  total_amount: number;
  payment_method: 'cash' | 'card' | 'check' | 'other';
  payment_reference?: string;
  status: 'completed' | 'voided' | 'refunded';
  transaction_date: string;
  notes?: string;
  lines: POSTransactionLine[];
  transaction_type_name?: string;
  customer_name?: string;
  cashier_name?: string;
}

export interface POSTransactionCreate {
  transaction_type_id: number;
  customer_id?: number;
  payment_method: 'cash' | 'card' | 'check' | 'other';
  payment_reference?: string;
  notes?: string;
  lines: Omit<POSTransactionLine, 'id'>[];
}

export interface POSCashMovement {
  id: number;
  company_id: number;
  session_id: number;
  movement_type: 'cash_in' | 'cash_out';
  amount: number;
  reason: string;
  reference?: string;
  created_at: string;
  created_by: number;
  created_by_name?: string;
}

export interface POSCashMovementCreate {
  movement_type: 'cash_in' | 'cash_out';
  amount: number;
  reason: string;
  reference?: string;
}

export interface CashierSalesReport {
  cashier_id: number;
  cashier_name: string;
  total_sales: number;
  total_returns: number;
  net_sales: number;
  transaction_count: number;
  average_transaction: number;
}

export interface InventorySalesReport {
  inventory_item_id: number;
  inventory_item_code: string;
  inventory_item_name: string;
  quantity_sold: number;
  total_sales: number;
  average_price: number;
}

export interface CartItem {
  item_id: number;
  description: string;
  quantity: number;
  unit_price: number;
  discount_percentage: number;
  discount_amount: number;
  tax_amount: number;
  line_total: number;
}

export interface InventoryItem {
  id: number;
  item_code: string;
  description: string;
  selling_price: number;
  cost_price: number;
  is_active: boolean;
}

// Additional types for compatibility with specification
export interface TillSession {
  id: number;
  tillId: number;
  userId: number;
  openingDate: string;
  closingDate?: string;
  openingBalance: number;
  expectedClosingBalance?: number;
  actualClosingBalance?: number;
  variance?: number;
  status: 'Open' | 'Closed' | 'Reconciled';
  reconciliationNotes?: string;
}

export interface POSPayment {
  id: number;
  paymentMethod: string;
  amount: number;
  referenceNumber?: string;
  paymentDetails?: any;
}

export interface ReceiptData {
  transaction: POSTransaction;
  companyInfo: any;
  tillInfo: any;
  cashierName: string;
  receiptHeader?: string;
  receiptFooter?: string;
}
